src/profiling/profiler.cpp

```diff
@@ -1,442 +1,401 @@
+class recorder {
+    // used to mark that the recorder is not currently timing a region.
+    static constexpr std::size_t npos = std::numeric_limits<std::size_t>::max();
+
+    // The index of the region being timed.
+    // If set to npos, no region is being timed.
+    std::size_t index_ = npos;
+
+    // The time at which the currently profiled region started.
+    time_point start_time_;
+
+    // One accumulator for call count and wall time for each region.
+    std::vector<profile_accumulator> accumulators_;
+
+public:
+    // Return a list of the accumulated call count and wall times for each region.
+    const std::vector<profile_accumulator>& accumulators() const;
+
+    // Start timing the region with index.
+    // Throws std::runtime_error if already timing a region.
+    void enter(std::size_t index);
+
+    // Stop timing the current region, and add the time taken to the accumulated time.
+    // Throws std::runtime_error if not currently timing a region.
+    void leave();
+
+    // Reset all of the accumulated call counts and times to zero.
+    void clear();
+};
+
+// Manages the thread-local recorders.
+class profiler {
+    std::vector<recorder> recorders_;
+
+    // Hash table that maps region names to a unique index.
+    // The regions are assigned consecutive indexes in the order that they are
+    // added to the profiler with calls to `index_from_name()`, with the first
+    // region numbered zero.
+    std::unordered_map<const char*, std::size_t> name_index_;
+
+    // The name of each region being recorded, with index stored in name_index_
+    // is used to index into region_names_.
+    std::vector<std::string> region_names_;
+
+    // Used to protect name_index_, which is shared between all threads.
+    std::mutex mutex_;
+    time_point tstart_;
+    time_point tstop_;
+    bool running_ = false;
+
+public:
+    profiler();
+
+    void start();
+    void stop();
+    void restart();
+    void enter(std::size_t index);
+    void enter(const char* name);
+    void leave();
+    const std::vector<std::string>& regions() const;
+    std::size_t index_from_name(const char* name);
+    profile results() const;
+};
+
+
+// Utility structure used to organise profiler regions into a tree for printing.
+struct profile_node {
+    static constexpr std::size_t npos = std::numeric_limits<std::size_t>::max();
+
+    std::string name;
+    double time = 0;
+    std::size_t count = npos;
+    std::vector<profile_node> children;
+
+    profile_node() = default;
+    profile_node(std::string n, double t, std::size_t c):
+        name(std::move(n)), time(t), count(c) {}
+    profile_node(std::string n):
+        name(std::move(n)), time(0), count(npos) {}
+};
 
-void profiler_node::print(std::ostream& stream, double threshold) {
-    // convert threshold from proportion to time
-    threshold *= value;
-    print_sub(stream, 0, threshold, value);
+namespace data {
+    arb::util::profiler profiler;
```

Comment:
I don't think we use this convention for global objects elsewhere. I've a personal preference for a `g_` prefix on the global variable in question, as it's very much in the programmer's face :) We can take this one to bikeshed.

Reply:
To the bikeshed!

Reply:
Added a static member `static profiler::get_global_profiler()` that returns a reference to a static `profiler`, similarly to the cthread thread pool implementation.l